### PR TITLE
[models] add ViTSTR TF and PT and update ViT to work as backbone

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- adds ViTSTR in TF and PT
- update ViT config
- add pos_interpolation to ViT to make it work also on higher image resolutions (case: recognition) as backbone
- ViT tf refactor patchify to reduce memory consuption (now it is similar to PT implementation TF: vit_s: 8.9 GB to PT: 7.5GB)
  (This difference is normal faiced the same % diff in SAR and MASTER between TF and PT)
- apply some fixes
- added model configs in this PR in both frameworks are: vit_s, vit_b (updated), viststr_small, vitstr_base 

Any feedback is very welcome :hugs: 

NOTE: 
Unlike the SAR or MASTER architecture, I am not able to fully train the model because ViT requires a lot of data and I cannot muster the computing power. So just a little test this time based on our word generator to show that it trains well.

slow tests: passed

PT: 

```
(doctr-dev) felix@felix-GS66-Stealth-11UH:~/Desktop/doctr$ python3 /home/felix/Desktop/doctr/references/recognition/train_pytorch.py vitstr
2022-09-19 09:35:14.038958: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Namespace(amp=False, arch='vitstr', batch_size=64, device=None, epochs=10, find_lr=False, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', input_size=32, lr=0.001, max_chars=12, min_chars=1, name=None, pretrained=False, push_to_hub=False, resume=None, sched='cosine', show_samples=False, test_only=False, train_path=None, train_samples=1000, val_path=None, val_samples=20, vocab='french', wb=False, weight_decay=0, workers=None)
Validation set loaded in 1.727s (2520 samples in 40 batches)
WARNING:root:Invalid model URL, using default initialization.
Train set loaded in 0.002008s (126000 samples in 1968 batches)
Validation loss decreased inf --> 3.90789: saving state...                                                                                                      
Epoch 1/10 - Validation loss: 3.90789 (Exact: 1.35% | Partial: 2.26%)
Validation loss decreased 3.90789 --> 3.54396: saving state...                                                                                                  
Epoch 2/10 - Validation loss: 3.54396 (Exact: 3.93% | Partial: 5.00%)
```

TF:

```
(doctr-dev-tf) felix@felix-GS66-Stealth-11UH:~/Desktop/doctr$ python3 /home/felix/Desktop/doctr/references/recognition/train_tensorflow.py vitstr
Namespace(amp=False, arch='vitstr', batch_size=64, epochs=10, find_lr=False, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', input_size=32, lr=0.001, max_chars=12, min_chars=1, name=None, pretrained=False, push_to_hub=False, resume=None, show_samples=False, test_only=False, train_path=None, train_samples=1000, val_path=None, val_samples=20, vocab='french', wb=False, workers=None)
Validation set loaded in 0.002643s (2520 samples in 40 batches)
WARNING:root:Invalid model URL, using default initialization.
Train set loaded in 0.004127s (126000 samples in 1968 batches)
Validation loss decreased inf --> 3.9933: saving state...                                                                                                       
Epoch 1/10 - Validation loss: 3.9933 (Exact: 3.45% | Partial: 3.73%)
Validation loss decreased 3.9933 --> 3.74347: saving state...                                                                                                   
Epoch 2/10 - Validation loss: 3.74347 (Exact: 7.50% | Partial: 7.78%)
```

Additional:
pred works also: (only tested with a model which reaches ~15% exact after 9 epochs trained with WordGenerator samples)
```
Word(value='฿฿_', confidence=0.048),
Word(value='4฿^฿', confidence=0.038),
Word(value='|', confidence=0.05),
Word(value='ërwW', confidence=0.031),
Word(value='x¢฿MMo', confidence=0.018),
```

### CODE REVIEW

File: doctr/models/recognition/vitstr/pytorch.py
Comment (felixdittrich92) [inline doctr/models/recognition/vitstr/pytorch.py:191]: @frgfm that's what i mean we want use the pretrained model but need to init it with different args.. i have not really an idea how we can reach this with IntermediateLayerGetter

File: doctr/models/recognition/vitstr/base.py
Comment (frgfm) [inline doctr/models/recognition/vitstr/base.py:57]:
```python
self._embedding = list(vocab) + ["<eos>", "<sos>", "<pad>"]
```

**Review by frgfm — COMMENTED** [review]
thanks Felix! I'd have to go through the paper more thoroughly for an accurate review, but I added some comments!

File: doctr/models/recognition/vitstr/pytorch.py
Comment (frgfm) [reply doctr/models/recognition/vitstr/pytorch.py]: well we can use `IntermediateLayerGetter` and take the last layer :)

**Comment by felixdittrich92** [comment]
> thanks Felix! I'd have to go through the paper more thoroughly for an accurate review, but I added some comments!

yeah i have still to fix the loss computation my main problem is how we can use ViT in this case as backbone .. it needs a completly different initialization ...

File: doctr/models/recognition/vitstr/pytorch.py
Comment (felixdittrich92) [reply doctr/models/recognition/vitstr/pytorch.py]: we have to init ViT with completly other args ... i dont know how this can work :sweat_smile:

Comment (felixdittrich92) [reply doctr/models/recognition/vitstr/pytorch.py]: ViT-B:

```
input_shape: Tuple[int, int, int] = (32, 32, 3),
patch_size: Tuple[int, int] = (4, 4),
d_model: int = 768,
num_layers: int = 12,
num_heads: int = 12,
ffd_ratio: int = 4,
dropout: float = 0.0,
num_classes: int = 1000,
include_top: bool = True,
```

We need:
```
input_shape: Tuple[int, int, int] = (32, 128, 3),
patch_size: Tuple[int, int] = (4, 8),
d_model: int = 384,
num_layers: int = 12,
num_heads: int = 6,
ffd_ratio: int = 4,
dropout: float = 0.0,
include_top: bool = False,
```

**Comment by felixdittrich92** [comment]
Found a way to make it work with IntermediateLayerGetter update will follow soon

**Comment by felixdittrich92** [comment]
done

File: doctr/models/modules/vision_transformer/pytorch.py
Comment (felixdittrich92) [reply doctr/models/modules/vision_transformer/pytorch.py:29]: Yep it's only a fix If we use ViT as backbone for recognition task otherwise you can easily specify the Input and patch size

**Review by odulcy-mindee — APPROVED** [review]
Looks good to me. Maybe I'll wait for a review from @frgfm for this model

**Comment by felixdittrich92** [comment]
@frgfm for TF we will switch later to IntermediateLayerGetter if we have pretrained ViT models :+1:

**Review by odulcy-mindee — APPROVED** [review]

**Comment by felixdittrich92** [comment]
@frgfm i will open another PR to add pretrained weights for ViT. We can add your requested changes into this (if there is anything :) )

File: doctr/models/modules/vision_transformer/pytorch.py
Comment (frgfm) [inline doctr/models/modules/vision_transformer/pytorch.py:29]: tricky condition: what are all the possible cases as input, and what do we want as patch_size for each?

**Review by frgfm — COMMENTED** [review]
Let's only set all the arguments for/against the patch_size default value (cf. comment)

File: doctr/models/modules/vision_transformer/pytorch.py
Comment (felixdittrich92) [reply doctr/models/modules/vision_transformer/pytorch.py:29]: oh it could be anything ..
currently in classification case: 32x32 (check)
recognition case: 32x128 (check)
detection case 1024x1024 (not handled)
any other size (not handled) 

it will not fail but each size needs a different patch_size

Comment (frgfm) [reply doctr/models/modules/vision_transformer/pytorch.py:29]: Ok two questions then:
- how should the scale impact the patch size ? (N,N) --> (H,W) implies that (2N,2N) --> (?,?)
- how should the aspect ratio impact the patch size? I see that (32,32) --> (4, 4), but why (32,128) doesn't do (4, 16) ?

Comment (felixdittrich92) [reply doctr/models/modules/vision_transformer/pytorch.py:29]: https://github.com/baudm/parseq/blob/main/configs/model/parseq.yaml
https://github.com/baudm/parseq/blob/main/configs/experiment/vitstr.yaml

**Comment by ghassenBtb** [comment]
> 
@felixdittrich92 Did you add pretrained weights for ViTSTR recognition model ? I looked in the code and there was no url link to pretrained weights.
Thanks

**Comment by felixdittrich92** [comment]
Hi @ghassenBtb :wave:,

no, i have added only pretrained weights for the ViT backbone. 
I have no access to mindee`s internal dataset which was used to train the recognition models so ping @charlesmindee @odulcy-mindee

**Comment by ghassenBtb** [comment]
Thank you Felix for your quick response :) 
@charlesmindee @odulcy-mindee did you pretrain ViTSTR on your internal dataset? 
Thanks

**Comment by charlesmindee** [comment]
Hi @ghassenBtb, we didn't pretrain ViTSTR on our internal dataset

**Comment by felixdittrich92** [comment]
Hi @charlesmindee i think the question from @ghassenBtb was more about if you can train the model internally and provide the checkpoints 😅

**Comment by ghassenBtb** [comment]
Yes it would be great if you can train the model on your internal french dataset :) 
It will certainly have superior recognition performance than CRNN-based recognition models as it will learn semantic and syntactic properties.

**Pull request merged**
